### PR TITLE
Allow to create Mainnet Filecoin accounts

### PR DESCRIPTION
**Pull request by spylogsster**
Resolves https://github.com/brave/brave-browser/issues/22947

Separated usage of Mainnet/Testnet addresses to use proper keys for the networks. Added creation of adresses based on current Filecoin network

https://user-images.githubusercontent.com/2965009/170084495-b7f4e146-8e85-43dc-8998-013831ceabb5.mov

https://github.com/brave/security/issues/870

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

- Select Filecoin Mainnet as active network
- Create Filecoin Mainnet accounts
- Switch to Filecoin Testnet network
- Check that you cannot see Mainnet accounts anymore and vice versa

### CODE REVIEW

File: components/brave_wallet/browser/filecoin_keyring.cc [hunk 1]
```diff
@@ -36,7 +36,13 @@ bool GetBLSPublicKey(const std::vector<uint8_t>& private_key,
 }
 }  // namespace
 
-FilecoinKeyring::FilecoinKeyring() = default;
+FilecoinKeyring::FilecoinKeyring(const std::string& chain_id) {
+  auto* network = (chain_id == brave_wallet::mojom::kFilecoinMainnet)
```
Comment (darkdh) [inline components/brave_wallet/browser/filecoin_keyring.cc:40]: mojom::kFilecoinMainnet and mojom::kFilecoinTestnet are already a string, there is no point getting their pointer and dereferencing it

File: browser/brave_wallet/filecoin_keyring_unittest.cc [hunk 2]
```diff
@@ -190,5 +190,4 @@ TEST(FilecoinKeyring, SignTransaction) {
   EXPECT_EQ(address, "t1lqarsh4nkg545ilaoqdsbtj4uofplt6sto26ziy");
   EXPECT_TRUE(keyring.SignTransaction(&transaction.value()));
 }
-
```
Comment (darkdh) [inline browser/brave_wallet/filecoin_keyring_unittest.cc:193]: Please remove this unrelated change.

File: components/brave_wallet/browser/filecoin_keyring.cc [hunk 3]
```diff
@@ -131,14 +137,29 @@ void FilecoinKeyring::RestoreFilecoinAccount(
   }
 }
 
+void FilecoinKeyring::ChainChangedEvent(const std::string& chain_id,
+                                        mojom::CoinType coin) {
+  if (coin != mojom::CoinType::FIL) {
+    return;
+  }
+  auto* network = (chain_id == brave_wallet::mojom::kFilecoinMainnet)
```
[the remark this answers is not in the document]
Comment (spylogsster) [reply components/brave_wallet/browser/filecoin_keyring.cc:145]: yes

File: browser/brave_wallet/filecoin_keyring_unittest.cc (hunk 2 above)
Comment (spylogsster) [reply browser/brave_wallet/filecoin_keyring_unittest.cc]: done

File: components/brave_wallet/browser/filecoin_keyring.cc (hunk 1 above)
Comment (spylogsster) [reply components/brave_wallet/browser/filecoin_keyring.cc]: done

File: components/brave_wallet/browser/filecoin_keyring.cc [hunk 4]
```diff
@@ -131,14 +137,29 @@ void FilecoinKeyring::RestoreFilecoinAccount(
...
+                      ? mojom::kFilecoinMainnet
+                      : mojom::kFilecoinTestnet;
+
+  if (network_prefix_ != std::string(network)) {
```
[the remark this answers is not in the document]
Comment (spylogsster) [reply components/brave_wallet/browser/filecoin_keyring.cc]: done

**Review by darkdh — APPROVED** [review]

**Review by kdenhartog — APPROVED** [review]
I spent some time reviewing the localhost network to see if there was an issue where a user who creates an account for localhost and then switches the localhost node from testnet to mainnet. From what I can tell as long as the user is properly entering the chain id when they're signing up then this won't be an issue, but if they do miss assign the chain id that could create a replay issue. Given we default the localhost network to testnet, I don't think it's likely for this issue to occur and so therefore am fine with this moving forward.

I did look into whether there was an API that could help us determine the network in use, but I couldn't find anything obvious. One method I thought of for this is to rely on the [https://lotus.filecoin.io/reference/api/state/#statenetworkname](stateNetworkName API) to get the name if that corresponds with mainnet or testnet, but from their docs it doesn't appear to be related. Alternatively we could hardcode the testnet and mainnet genesis block hashes in and then query the node for the genesis block to check the network, but I didn't see any APIs that made this available.

So for these reasons I think it's fine for us to rely on the user putting the correct chain_id in when attempting to override the localhost node.

**Comment by srirambv** [comment]
Verification passed on 

Brave | 1.41.20 Chromium: 102.0.5005.61 (Official Build) nightly (64-bit)
-- | --
Revision | `0e59bcc00cc4985ce39ad31c150065f159d95ad3-refs/branch-heads/5005@{#819}`
OS | Windows 11 Version 22H2 (Build 22621.1)

- Verified test plan from PR description
- Verified selecting Filecoin Mainnet only shows mainnet account and hides testnet accounts
- Verified selecting Filecoin Testnet only shows test net accounts and hides mainnet account

https://user-images.githubusercontent.com/17010094/170912556-410b9dcc-86e0-402b-b9ac-f5ee60968353.mov
